### ISSUE TITLE
Collapsed JSP output at specific conditions

### ISSUE THREAD

**Issue by c-yan** [javaee/glassfish#21862]
## Environment Details
* GlassFish Version (and build number): 4.1.2
* JDK version: Oracle JDK 1.8.0_131
* OS: Windows Server 2016
* Database: none

----------

## Attached File
[gf41-test.zip](https://github.com/javaee/glassfish/files/1079823/gf41-test.zip)

----------

## Problem Description
The JSP output seems to be collapsed at specific conditions and all embedded scriptlets of JSP are collectively output in the end.
Attached is a code that reproduces the incident and the output result is as below:

expected result:
```
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>test</title>
    </head>
    <body>
      <p>Hello world!</p>
    </body>
  </html>

```
actual result:
```
  <!DOCTYPE html>
  <html lang="">
    <head>
      <meta charset="">
      <title>test</title>
    </head>
    <body>
      <p> world!</p>
    </body>
  </html>
  
  enutf-8Hello

```
## Steps to reproduce
1. Deploy target/gf41-test.war in the attached gf41-test.zip.
2. Try access to /gf41-test/test.

## Impact of Issue
Eclipse BIRT Runtime is not working.
Ref: https://bugs.eclipse.org/bugs/show_bug.cgi?id=459440
It was properly working in Glassfish 3.1.

**Comment by ruolli** [comment]
The issue happens when 
```
<%@ page session="false" buffer="none" %>
```

**Comment by ruolli** [comment]
The change I am going to submit is like 
```
diff --git a/impl/src/main/java/org/apache/jasper/compiler/Generator.java b/impl/src/main/java/org/apache/jasper/compiler/Generator.java
index 87a9b26..2988acd 100644
--- a/impl/src/main/java/org/apache/jasper/compiler/Generator.java
+++ b/impl/src/main/java/org/apache/jasper/compiler/Generator.java
@@ -856,8 +856,15 @@ class Generator {
         }
 
         public void visit(Node.Expression n) throws JasperException {
+
             n.setBeginJavaLine(out.getJavaLine());
-            out.printil("out.print(" + n.getText() + ");");
+            if (genBytes) {
+              out.printil("((org.apache.jasper.runtime.JspWriterImpl)out).write( (byte[])" 
+                          + n.getText() + ".getBytes(_jspx_encoding)," + " 0, ((byte[])" 
+                          + n.getText() + ".getBytes(_jspx_encoding)).length );");
+            } else {
+              out.printil("out.print(" + n.getText() + ");");
+            } 
             n.setEndJavaLine(out.getJavaLine());
         }

```

**Comment by ruolli** [comment]
As to why GlassFish 3.1 does not have the issue, I checked the jsp compiler code. 
I don't see any obvious change which might cause the difference. 
So, I am guessing the reason lies in 3rd party packages GlassFish 3.1 and GlassFish 4.1 are using. 
Not confirmed yet.

**Comment by rlubke** [comment]
See:  https://github.com/javaee/grizzly/issues/1839  

This should be resolved in GF5.

**Comment by ruolli** [comment]
As Ryan said, GF 5 does not have the issue, since Grizzly has fixed the buffer issue. 
So, it's not necessary to submit the previous change I mentioned.

I am going to close this issue as "Not a Bug".

**Comment by toshihiro-shiino** [comment]
GF 4.1.2 is affected by this bug, is this ticket "Not a Bug"?

**Comment by ruolli** [comment]
@toshihiro-shiino Or, we should integrate the latest Grizzly to GF 4.1 to fix the issue, instead of submitting the previous change to GF 4.1 .

**Comment by toshihiro-shiino** [comment]
I understood it as the following but is this correct?

Grizzly corrected this bug in June of last year.
Fixed release is available as 2.3.26 in July of last year.
Since GF 4.1.2 did not update Grizzly, the version of Grizzly included in GF 4.1.2 is 2.3.23, and there is this bug in GF 4.1.2.

**Comment by rlubke** [comment]
It's still an outstanding issue that the sustaining teams for GF 4.x should address.  They should uptake the new version of Grizzly 2.3.x.

**Comment by ruolli** [comment]
@toshihiro-shiino Correct.

**Comment by toshihiro-shiino** [comment]
@ruolli thank you.

**Comment by edbratt** [comment]
Fixes are always provided to the master branch. We don't "patch" old open-source versions and there is no product support available for GlassFish 4.x. You are welcome to apply this change to your copy of GlassFish 4.1.2.

**Comment by toshihiro-shiino** [comment]
@edbratt 
Is my understanding as below correct?

1. The fix for this bug will be provided within GF5.0 and this ticket shall be closed as a bug.
2. The fix release will not be delivered for GF4.1.2, so we need to create one if we need it.

\# What I have concerned is expecting future GF5.0 release with fix implemented for a bug reproducible for the current official release of GF4.1.2 does not justify closing this ticket as "Not a Bug".

**Comment by toshihiro-shiino** [comment]
I've got the fixed module!!

```
$ git diff
diff --git a/nucleus/pom.xml b/nucleus/pom.xml
index 35060b4..cbc627e 100644
--- a/nucleus/pom.xml
+++ b/nucleus/pom.xml
@@ -145,7 +145,7 @@
         <stax-api.version>1.0-2</stax-api.version>
         <management-api.version>1.1-rev-1</management-api.version>
         <servlet-api.version>3.1.0</servlet-api.version>
-        <grizzly.version>2.3.23</grizzly.version>
+        <grizzly.version>2.3.26</grizzly.version>
         <saaj-api.version>1.3.4</saaj-api.version>
         <hk2.version>2.4.0-b31</hk2.version>
         <hk2.plugin.version>2.4.0-b31</hk2.plugin.version>
```
[nucleus-grizzly-all.zip](https://github.com/javaee/glassfish/files/1172574/nucleus-grizzly-all.zip)

**Comment by toshihiro-shiino** [comment]
I've verified fix with birt and attached test.

**Comment by ruolli** [comment]
This issue has been fixed in GF 5.0, 
and, we won't provide fix delivery for former GF releases. 
So, close this issue for now.
